### ISSUE TITLE
Issue targeting nodegroups - Invalid compound target: ( L@ ... )

### ISSUE THREAD

**Issue by thecosmicfrog** [saltstack/salt#26107]
Seeing a strange bug when targeting nodegroups. Some of my minion return fine, whereas others give me either "Minion did not return. [No response]" or "Minion did not return. [Not connected]".

As below:
```yaml
=== /etc/salt/master ===
nodegroups:
   non-critical: 'L@ansible-master,server1,server2,server3,server4,server5,salt-master,server6,server7,server8'
```
```console
=== Targeting from command line ===
root@salt-master:~# salt -N 'non-critical' test.version
server5:
    2015.5.3
salt-master:
    2015.5.3
server4:
    2015.5.2
server1:
    2015.5.3
server7:
    2015.5.2
server2:
    Minion did not return. [No response]
server3:
    Minion did not return. [No response]
ansible-master:
    Minion did not return. [No response]
server6:
    Minion did not return. [Not connected]
server8:
    Minion did not return. [No response]
```
```
=== Minion log from failing server (server2) ===
=== /var/log/salt/minion ===
2015-08-07 13:53:08,938 [salt.minion      ][ERROR   ] Invalid compound target: ( L@ansible-master,server1,server2,server3,server4,server5,salt-master,server6,server7,server8 )
```

Why would this be an "invalid compound target"?

Running `salt '*' <whatever>` works as expected. It's just nodegroups that have the issue.

**Comment by jfindlay** [comment]
@thecosmicfrog, that looks like a bug to me.  Thanks for the report.

**Comment by thecosmicfrog** [comment]
No bother at all. Always happy to report a bug. Please let me know if there's anything I can do to test further.

**Comment by vr-jack** [comment]
@thecosmicfrog Can you verify master version and the version of the non-working minions?

Might also check https://github.com/saltstack/salt/issues/24660. Try the commandline compound method and see if that works for you as well.

**Comment by thecosmicfrog** [comment]
@vr-jack Sure thing. See below:

=== Master ===
* OS: Debian 7 wheezy<br>
salt-master: 2015.5.3+ds-1~bpo70+2


=== Working Minions ===
* OS: Debian 8 jessie<br>
salt-minion: 2015.5.3+ds-1~bpo70+2<br>
* OS: Debian 8 jessie<br>
salt-minion: 2015.5.2+ds-1~bpo8+1
* OS: Debian 8 jessie<br>
salt-minion: 2015.5.2+ds-1~bpo8+1
* OS: CentOS 6<br>
salt-minion: 2015.5.3-3.el6
* OS: CentOS 7<br>
salt-minion: 2015.5.3-4.el7


=== Non-Working Minions ===
* OS: Debian 8 jessie<br>
salt-minion: 2014.1.13+ds-3<br>
Status: Minion did not return. [Not connected]
* OS: Debian 8 jessie<br>
salt-minion: 2014.1.13+ds-3<br>
Status: Minion did not return. [No response]
* OS: Debian 8 jessie<br>
salt-minion: 2014.1.13+ds-3<br>
Status: Minion did not return. [No response]
* OS: openSUSE 13.2<br>
salt-minion: 0.16.4-2.1.3<br>
Status: Minion did not return. [No response]
* OS: Ubuntu 14.04 LTS<br>
salt-minion: 0.17.5+ds-1<br>
Status: Minion did not return. [No response]


So, the correlation here is obvious, at least for the Debian systems. I'm running Minions from 2014. After checking, these are coming directly from the jessie repos, whereas the "working" salt-minion packages are coming from **debian.saltstack.com**.

In a lovely twist of irony, I may need to go and re-work my Salt states to ensure my dependencies are set up correctly to pull salt-minion from the correct repo.

Still, this bug is likely still of interest to the SaltStack developers.

**Comment by thecosmicfrog** [comment]
Also - note the differing status value between two Debian 8 systems running the **same** version of salt-minion, i.e.

*Minion did not return. [Not connected]*

and

*Minion did not return. [No response]*

**Comment by vr-jack** [comment]
@thecosmicfrog Thanks for the info. It's definitely a duplicate of https://github.com/saltstack/salt/issues/24660

@basepi Do we want to try and fix this backwards compatibility issue?

**Comment by thecosmicfrog** [comment]
@vr-jack It definitely seems to be. Hopefully I didn't waste too much of your time with this. Thanks for all your help and input.

**Comment by vr-jack** [comment]
@thecosmicfrog Nah. Was just in a bug fixing mood and trying to find something my level to fix. You should be able to salt -C to target them manually if you are looking at upgrading the older versions. I'll research the problem, though.

**Comment by vr-jack** [comment]
Found where older versions are failing. https://github.com/saltstack/salt/blob/2014.1/salt/minion.py#L1922 was altered in 2014.7 to handle  '(' and ')' as seen on this line https://github.com/saltstack/salt/blob/2014.7/salt/minion.py#L2571

I have broken my 2014.7 by altering that single line to get the results of the bug. If approval given, we can try to support that problem when sending to older clients.

**Comment by vr-jack** [comment]
@basepi @jfindlay 

Here is the start of the break:
https://github.com/saltstack/salt/commit/1c3536fb8e1b1814f6ba988efbb018bc98b7066f

This is the line in current that adds the (). https://github.com/saltstack/salt/blob/2015.5/salt/utils/minions.py#L103

In 2015.8 and dev, it was rewritten again. Here is the offensive lines: https://github.com/saltstack/salt/blob/2015.8/salt/utils/minions.py#L141-L143

Here is the test I groups I ran:

nodegroups:
  group1: 'host1'
  group2: 'host2'
  test1: 'N@group1 or N@group2'
  test2: 'host1 or N@group2'
  test3: 'L@host1,and,host2'
  test4: 'L@host1,host2'

All 4 tests pinged both hosts. Host1 is 2014.7 modified to break like 2014.1 (parens kill it). Host2 is 2015.5.3 unmodified. Master is 2015.5.3 modified to remove the (). To the best of my knowledge, the parenthesis are not needed and *should* not break anything, but it's a judgement call. I'll write the patches and do the 2 pull requests if approved.

**Comment by jfindlay** [comment]
@vr-jack, @thecosmicfrog, thanks for working on this issue.  @basepi is going to be the best judge of what to do here.

**Comment by basepi** [comment]
I fixed backwards compat between 2015.8 masters and 2015.5 and 2014.7 minions in #28148 for nodegroup matches.

Here are my test results. I spun up a minion for 2015.8.1, 2015.5.6, and 2014.7.6, and then tried different master versions controlling them.

- 2014.7.6 master works great with all three, including nodegroups.
- 2015.5.6 master works great with all three, including nodegroups.
- 2015.8.2 master works great with all three, including nodegroups. (but not nested nodegroups)

That will be true once 2015.8.2 is released with the above fix. 2015.8.1 masters don't play well with 2015.5 and 2014.7 minions when it comes to nodegroups.

Once the above is merged, I plan to close this issue.

**Comment by thecosmicfrog** [comment]
Thanks @basepi!

**Comment by ahammond** [comment]
Awesome! I can see us making the jump from trusty ole 2015.5.3 to 2015.8.2.
:)
